swap/src/cli/cancel_and_refund.rs

```diff
@@ -55,9 +56,27 @@ pub async fn cancel(
 
     tracing::info!(%swap_id, "Manually cancelling swap");
 
-    let (txid, subscription) = match state6.submit_tx_cancel(bitcoin_wallet.as_ref()).await {
-        Ok(txid) => txid,
+    match state6.submit_tx_cancel(bitcoin_wallet.as_ref()).await {
+        Ok((txid, _)) => {
+            let state = BobState::BtcCancelled(state6);
+            db.insert_latest_state(swap_id, state.clone().into())
+                .await?;
+            Ok((txid, state))
+        }
         Err(err) => {
+            if let Ok(tx) = state6.check_for_tx_cancel(bitcoin_wallet.as_ref()).await {
+                // Alice already cancelled swap, so Bob sets his state to BtcCancelled.
+                let state = BobState::BtcCancelled(state6);
+                db.insert_latest_state(swap_id, state.clone().into())
+                    .await?;
+                tracing::info!("Cancel transaction has already been confirmed on chain");
+                return Ok((tx.txid(), state));
+            }
+            if let ExpiredTimelocks::None { .. } =
+                state6.expired_timelock(bitcoin_wallet.as_ref()).await?
+            {
+                bail!("Cancel timelock hasn't expired yet");
+            }
             if let Ok(error_code) = parse_rpc_error_code(&err) {
                 tracing::debug!(%error_code, "parse rpc error");
                 if error_code == i64::from(RpcErrorCode::RpcVerifyAlreadyInChain) {
```

Comment:
The case below should never happen but if it does we should transition to BtcCancelled like above. If possible avoid duplicating the code above here but combine the two somehow

Reply:
As far as I know, or operation isn't supported with if let.